marconi-tutorial/src/Marconi/Tutorial/Env.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Marconi.Tutorial.Env where
+
+import Cardano.BM.Trace (Trace)
+import Control.Concurrent (MVar)
+import Control.Concurrent.MVar (newEmptyMVar)
+import Control.Lens (makeLenses)
+import Data.IORef (IORef, newIORef)
+import Data.Text (Text)
+import Marconi.Tutorial.CLI qualified as CLI
+import Marconi.Tutorial.Indexers.AddressCount (AddressCountIndexer)
+
+data Env = Env
+  { _envIndexers :: IndexersEnv
+  -- ^ Access the Tutorial indexers for querying purposes.
+  , _envCliArgs :: !CLI.Options
+  , _envStdoutTrace :: !(Trace IO Text)
+  }
+
+-- | Should contain all the indexers required by the tutorial.
+newtype IndexersEnv = IndexersEnv
+  { _addressCountIndexerEnv :: AddressCountIndexerEnv
+  }
+
+newtype AddressCountIndexerEnv = AddressCountIndexerEnv
+  { _addressCountIndexerVar :: IORef (MVar AddressCountIndexer)
```

Comment:
Why an `IORef` of an `MVar` instead of just an `MVar` or just an `IORef` (although, we may have harcoded the `MVar` somewhere in the codebase and moving to an `IORef` might not be suitable.)

Reply:
That's because in `Indexers` module, we need to share the indexer's `MVar` with the query thread. The issue is that the indexer thread already creates the MVar, and the way I found to share that `MVar` with the query thread is to use an `IORef` where I can just mutate the `MVar` once made available by the indexer thread.

Haven't thought too deeply about it, but if you had an alternative approach, please tell.

Reply:
Ah, Neil proposed the solution to first create the indexers, and then pass the `MVar` to the two threads. Seems like a good approach.

Reply:
I totally forgot where the http server should be plugged but yes, as we use them in both threads, it makes sense to create the `MVar` before we fork the http server.